### ISSUE TITLE
[ses] Convert SES SMTP credentials with Typescript

### ISSUE THREAD

**Issue by Cloudrage** [aws/aws-cdk#8937]
Bug (?) when creating/converting SES SMTP credentials with Typescript.

### Reproduction Steps

I've a problem while getting the **SecretKey** from `CfnAccessKey` in my "convertion" snippet : 
_const key = CfnAccessKeyUserSmtp.getAtt('SecretAccessKey').toString();_ 

```
// AccessKey
    const CfnAccessKeyUserSmtp =  new iam.CfnAccessKey(this, 'CfnAccessKeyUserSmtp', {userName: UserSmtp.userName});
// Convert to SMTP base64 password
    const crypto = require('crypto');
    const utf8 = require('utf8');
    const key = CfnAccessKeyUserSmtp.getAtt('SecretAccessKey').toString();
    const region = AppAccountRegion;

    const date = '11111111';
    const service = 'ses';
    const terminal = 'aws4_request';
    const message = 'SendRawEmail';
    const versionInBytes = [0x04];
    
    function sign(key: { map: (arg0: (a: any) => any) => ArrayBuffer | SharedArrayBuffer; }, msg: string) {
      return crypto.createHmac('sha256', Buffer.from(key.map((a: string) => a.charCodeAt(0))))
        .update(utf8.encode(msg))
        .digest('latin1')
        .split('');
    }
    
    let signature = sign((utf8.encode('AWS4' + key)).split(''), date);
    signature = sign(signature, region);
    signature = sign(signature, service);
    signature = sign(signature, terminal);
    signature = sign(signature, message);
    
    const signatureAndVersion = versionInBytes.slice(); //copy of array
    
    signature.forEach((a: string) => signatureAndVersion.push(a.charCodeAt(0)));
    
    const smtpPassword = Buffer.from(signatureAndVersion).toString('base64');
```
It seems that the value can't be retrieved correctly.
And that way I've always a wrong password converted.

But when I'm creating a Secret, the real value is perfectly added : 
```
// Secret Manager
    const SecretsManagerUserSmtp = new secretsmanager.Secret(this, 'SecretsManagerUserSmtp', {
      description: 'Secrets for '+AppName+'-'+AppEnv,
      encryptionKey: KmsKeySecretsManager,
      secretName: UserSmtpSecretsName,
      generateSecretString: {
        excludeUppercase: false,
        requireEachIncludedType: true,
        includeSpace: false,
        passwordLength: 20,
        excludePunctuation: true,
        excludeLowercase: false,
        excludeNumbers: false,
        secretStringTemplate: JSON.stringify({
          host: 'email-smtp.eu-west-1.amazonaws.com',
          port: '25',
          domain: R53DomainSes,
          username: CfnAccessKeyUserSmtp.ref,
          secretkey : CfnAccessKeyUserSmtp.getAtt('SecretAccessKey').toString(),
          password: smtpPassword
        }),
        generateStringKey: 'WaitingForSecretStringPropsInCDK'
      },
    });
    SecretsManagerUserSmtp.denyAccountRootDelete();
```
Also, when I'm creating a SingletonFunction to do the Job it's working as well.

Any ideas ?

### Environment

  - **CLI Version      : CDK 1.48.0 (build 6080fa8)**
  - **Framework Version: 6.14.5**
  - **Node.js Version: v12.15.0** 
  - **OS               : Linux**
  - **Language: TypeScript** 

--- 

This is :bug: Bug Report

**Comment by iliapolo** [comment]
Hi @Cloudrage 

I'm not sure I follow, but i'd like to point out one thing that might be related. I see you are using:

```ts
const key = CfnAccessKeyUserSmtp.getAtt('SecretAccessKey').toString();
```

The value of this attribute is only known at deploy time, so when you run the code locally during synth, it is actually a token.
You mentioned that when its executed inside a Lambda function it works - right? Thats because the lambda function is given the resolved value at deploy time.

I hope this helps. 

If not, can you please elaborate exactly what are you trying to do? 

> It seems that the value can't be retrieved correctly.

What do you mean by that? 

> But when I'm creating a Secret, the real value is perfectly added :

Again, not sure I follow...

Are you trying to generate a password using the above code? Or retrieve a generated one?

**Comment by Cloudrage** [comment]
Hello @iliapolo  !

Thank you for your reply.
Indeed, I'm trying to convert the SecretKey of a dedicated SMTP User inside the CDK Stack.
(Finally put User credentials (Access/Secret Keys) in the Secret, "openssl enc -base64" ready).

I have understood that the value is retrieved at deploy time, but why it can't be retrieved (or it's wrong) in my Convert snippet ?
The Secretsmanager snippet create the **secretkey** key with the correct **SecretKey** value from the User credentials.
In the end, the convert Snippet output me a wrong password.

It's overkill to do that with a Lambda (singleton + CustomResource).

**Comment by iliapolo** [comment]
@Cloudrage I'm still not a 100% following you. 

Lets take it step by step, first, Is this what you are trying to do:

1. Create an Access Key (Using `iam.CfnAccessKey`).
2. Use the value of the access key to generate a password. (Using your conversion snipper).
3. Store that password in secrets manager.

?

**Comment by Cloudrage** [comment]
Exactly !

**Comment by iliapolo** [comment]
@Cloudrage Ok, so:

### 1) Create an AccessKey

No problem, this works well.  

### 2) Use the value of the access key to generate a password. (Using your conversion snipper).

This cannot be executed during synth (i.e using normal code in your stack) because the value of the access key is only available **after** it is created (i.e deployed). This is because it is created by CloudFormation, not by CDK. 

So, if you want access to the actual value, you must run it inside a lambda.

### 3) Store that password in secrets manager.

Notice that secrets manager is normally used to generate a secret on its own, this is why you configure these rules: 

```ts
        includeSpace: false,
        passwordLength: 20,
        excludePunctuation: true,
        excludeLowercase: false,
        excludeNumbers: false,
```

But here, you are using the secret manager to store a secret that you generate on your own, which means all these properties have no affect, and essentially you are bypassing the main purpose of it.

Can I ask, why are you generating the password on your own? is it because you need to comply with certain password restrictions that secret manager doesn't support? Also, why are you using the value of the access key in your generated password? Is this some standard practice? 

To summarize, if you absolutely have to both use the actual value of the access key, and generate the password on your own, than you have to do it within a lambda. Also, if you think about it, its way more secure, since you won't be writing secret values in your template, they will only be calculated at deploy time on the CloudFormation side.

**Comment by Cloudrage** [comment]
@iliapolo , since few years, I use to create those credentials with Sceptre (Troposphere) with a Hook converting SMTP password to be used by our app. These app are getting credentials/values from SecretManager.
Like you've said, this is done after creating/update Cfn resource.

I'm actually evaluating CDK to eventually migrate from Sceptre, on it.
So in that case, I have to write another part of the code to make that Hook with a Lambda here, it seems.

For SecretManager, like I've wrote in the snippet : "**generateStringKey: 'WaitingForSecretStringPropsInCDK'**"
See #5810, we can't actually create String only.
The "generateStringKey" is here for nothing, waiting for that _SecretString_ Prop.


So, you confirmed the Lambda solution (as Custom Resource) for that use case.
Thanks again.

**Comment by iliapolo** [comment]
@Cloudrage I understand. Thanks for the clarification 👍
